### PR TITLE
[distributed] Add cpp-httplib to pytorch

### PR DESCRIPTION
**Pull request by PaliC**
Adds https://github.com/yhirose/cpp-httplib such that we are able to use https for host to host communication in distributed (specifically torchrun)

Todo: We likely need to add cpp-httplib somewhere in the build (cmake/bazel) but first we should write the code for it.

### CODE REVIEW

File: third_party/cpp-httplib
Comment (d4l3k) [inline third_party/cpp-httplib:1]: Should we be pinning it to the latest release rather than master? https://github.com/yhirose/cpp-httplib/tree/v0.15.3

**Review by d4l3k — APPROVED** [review]
LGTM -- this looks good to me :)

It's MIT so shouldn't have any licensing concerns. Adding it to the build should be easy since it's just a single header file

**Review by Skylion007 — APPROVED** [review]

**Comment by jeffdaily** [comment]
I don't know how this is working.  I just cloned tip of main with this PR merged and I can't run `git submodule update --init --recursive` or perform a recursive clone.  I'm using the CI image

ghcr.io/pytorch/ci-image:pytorch-linux-focal-rocm-n-py3-6cae374c60eb8cbdc0e9d001d8ae4e21642e3f54

I'm getting the error:

```bash
~/pytorch$ git submodule update --init --recursive
Cloning into '/var/lib/jenkins/pytorch/third_party/cpp-httplib'...
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

**Comment by jeffdaily** [comment]
Is there a reason why we used git@github.com instead of https?  Can we change it to https?  That fixes the problem for me.

**Comment by d4l3k** [comment]
@jeffdaily good point -- I'll update this

**Pull request closed without merging**